File: Kth to Last Element (Linked List)/kthFromLast.java
public class kthFromLast {

	public static void main(String[] args) {

		//An example of how it works
		Node a = new Node(10);
		a.add(9);
		a.add(8);
		a.add(7);
		a.add(6);
		a.add(5);
		a.add(4);
		a.add(3);
		a.add(2);
		a.add(1);
		a.add(0);

		Node b = findKthFromLast(a, 4); //expect 4
		System.out.println(b.getData());

	}

	public static Node findKthFromLast(Node n, int k) {

		if (n == null) {
			System.out.println("Invalid n.");
			return null;
		} else if (k > n.getNodeCount() || k < 0 || n == null) {
			System.out.println("Invalid k.");
			return null;
		}

		Node first = n;
		Node second = n;

		for (int i = 0; i < k; i++) {
			second = second.next;
		}

		while (second.next != null) {
			first = first.next;
			second = second.next;
		}

		return first;

	}

	public static class Node {

		Node next;
		private int data;
		private int nodeCount;

		public Node(int data) {
			this.data = data;
			this.nodeCount = 0;
		}

		public void add(int data) {
			Node temp = new Node(data);
			Node current = this;
			while (current.next != null) {
				current = current.next;
			}
			current.next = temp;
			nodeCount++;
		}

		public int getNodeCount() {
			return nodeCount;
		}

		public int getData() {
			return data;
		}
	}

}

File: Kth to Last Element (Linked List)/kth_from_last.rb
class Node

  attr_accessor :next
  attr_reader :data, :node_count

  def initialize(data)
    @data = data
    @node_count = 0
    @next = nil
  end

  def add(data)
    temp = Node.new(data)
    current = self
    while current.next != nil
      current = current.next
    end
    current.next = temp
    @node_count += 1
  end

end

def find_kth_from_last(node, k)
  #handling edge cases
  if node == nil
    puts "Invalid node"
    return nil
  elsif !k || k > node.node_count || k < 0
    puts "Invalid k"
    return nil
  end

  first = node
  second = node

  k.times { second = second.next }

  while second.next
    first = first.next
    second = second.next
  end

  first

end

n = Node.new(10)
n.add(9)
n.add(8)
n.add(7)
n.add(6)
n.add(5)
n.add(4)
n.add(3)
n.add(2)
n.add(1)
n.add(0)
kth_to_last = find_kth_from_last(n, 4)
p kth_to_last.data #expect 4
kth_to_last = find_kth_from_last(n, 2)
p kth_to_last.data #expect 2